<!DOCTYPE html>
<html lang="en"> 
<head>
  <meta charset='utf-8'>

  <title>RECOT: Read Coordinate Transformer</title>

  <style type="text/css">
        /* ------------------------------------------------------------------------------------------------------------- RESET */

        html, body, div, form, p,
        code, pre { margin: 0; padding: 0; border: 0; vertical-align: baseline; }

        /* ------------------------------------------------------------------------------------------------------------- TEMPLATE */

        body { font: 14px/1.333 sans-serif; color: #444; background: #eee; }

        a { color: #980905; }
        a:hover, a:focus, a:active { text-decoration: none; }

        h1 { margin: 0 0 0.5em; font-size: 24px; }
        h2 { margin: 0 0 0.75em; font-size: 18px; }
        h3 { margin: 0 0 0.333em; font-size: 16px; }
        p { margin: 0 0 1.333em; }
        em { font-style: italic; }
        table { border-collapse: separate; border-spacing: 0; margin: 0; vertical-align: middle; }
        th { font-weight: bold; }
        th, td { padding: 5px 25px 5px 5px; text-align: left; vertical-align: middle; }
        pre, code { font-family: monospace, sans-serif; font-size: 1em; color:#080; }

        /* ------------------------------------------------------------------------------------------------------------- TEMPLATE */

        .container { position:relative; overflow:hidden; width: 600px; padding: 40px 60px; border: 1px solid #ccc; margin: 40px auto 20px; background: #fff; -webkit-box-shadow: 0 0 15px rgba(0,0,0,0.1); -moz-box-shadow: 0 0 15px rgba(0,0,0,0.1); box-shadow: 0 0 15px rgba(0,0,0,0.1); }

        .container pre,
        .container .prettyprint { padding: 0; border: 0; margin: 0 0 20px; font-size: 13px; background: #fff; }

        .ribbon { position: absolute; top: -1px; right: -1px; opacity: 0.9; }
        .ribbon:hover, .ribbon:focus, .ribbon:active { opacity: 1; }
        .ribbon img { display: block; border: 0; }

        .header { padding-right:80px; }

        .section { margin: 40px 0 20px; }

        .example { padding: 20px; border: 1px solid #ccc; margin: 10px -20px 20px; }

        .footer { margin: 20px 0 50px; font-size: 11px; color: #666; text-align: center; }
        .footer a { color: #666; }
  </style>
  <!-- import from css3 github buttons on 
       http://nicolasgallagher.com/lab/css3-github-buttons/ -->
  <link rel="stylesheet" href="gh-buttons.css">

  <!-- prettyify -->
  <link rel="stylesheet" href="http://google-code-prettify.googlecode.com/svn/trunk/src/prettify.css"> 
  <script src="http://google-code-prettify.googlecode.com/svn/trunk/src/prettify.js"></script> 
</head>
<body onload="prettyPrint()">
  <a href="http://github.com/sesejun/recot"><img style="position: absolute; top: 0; right: 0; border: 0;" src="http://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub" /></a>

  <div class="container">

    <div class="header">
      <h1><a href="http://github.com/sesejun/recot">RECOT: Read Coordinate Transformer</a></h1>
      RECOT transforms coordinate of next-generation sequencer reads from your genome/gene to other species genome. RECOT is useful to compare RNA-seq, ChIP-seq and CLIP-seq sequences between closely-related species.
    </div>
    <img alt="screen shot of RECOT result" src="images/screenshot.png" align=right />
    
    
    <div class="section">
      <h2>Requirements (Data)</h2>
      <ul>
	<li>Reads of your original species (1)</li>
	<li>Two different genomes (Your originally experimented species and target species) (2,4)</li>
	<li>Gene regions of your original genome in GFF format (3)</li>
	<li>(Option) Relationships of genes between two species</li>
      </ul>
      <p>
      <img alt="Illustration of Input Files" src="images/inputfiles.png" width=589 />
      </p>
    </div>
    <div class="section">
      <h2>Requirements (Software)</h2>
      <ul>
	<li>Python (>=2.6)</li>
	<li>Short read alignment program, such as <a href="http://bio-bwa.sourceforge.net/">BWA</a> and <a href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie</a></li>
	<li>Alignment program for long sequence, such as <a href="http://research-pub.gene.com/gmap/">GMAP/GSNAP</a></li>
	<ul>
	  <li>Both types of alignment programs must support SAM output because our programs use the format</li>
	</ul>
      </ul>
    </div>

    <div class="section">
      <h2>Simple Usage</h2>
      <p>As example, we here transform coordinate of ChIP-seq reads on <i>D. simulans</i> to <i>D. melanogaster</i> genome. 
	"sample" directory contains chromosome 2R of both <i>D. simulans</i> and <i>D. melanogaster</i>, and SAM file contains ChIP-seq result mapped onto <i>D. simulans</i> genome.
	Scripts will suggest you the next action.
	By changing settings in configuration file "settings.ini", 
	you can change both input and output file names.
      </p>
      <h3>Install [First time only] </h3>
      <p>If you have Git, you can clone latest version by running:
	<pre>
	  git clone git://github.com/sesejun/recot
	</pre>
      Don't worry when you do not have Git! Download the <a href="http://github.com/sesejun/recot/zipball/master">zipped codes</a> and extract it.
      </p>

      <h3>Extract gene sequences</h3>
      <p>First, extract gene sequences. "recot_extract.py" will generate "sample/dsim_geneseq.fasta" which contains gene and its regulatory regions (upstream and downstream 500bp as default).
	<pre>
	  python recot_extract.py -c settings_sample.ini
	</pre>
      </p>
      <img alt="Extract Gene Region" src="images/extract.png" width=602 />
      
      <h3>Align the gene sequences against target genome</h3>
      <p>Map the sequences in "data/dsim_geneseq.fasta" onto target genome. 
	We here use GMAP to map the sequences.
	<pre>
	  gmap_build -d dme sample/dmel-2R.fasta
	  gmap -f samse -d dme sample/dsim_geneseq.fasta > sample/dsim_geneseq_on_dme.sam
	</pre>
	By changing gmap option "samse" to "sampe", it can handle paired end sequence.
      Then, you can select the best matched sequences from the SAM file.
	<pre>python recot_rm_overlap.py -c settings_sample.ini</pre>
      </p>
      <p>
      <img alt="Alignment" src="images/align.png" />
      </p>

      <h3>Map short reads against target genes/genome</h3>
      <p>Map your reads onto original genome sequence (In the running example, <i>D. simulans</i>). Mapped result file in this sample is "data/ERR020078-2R-1M.sam", 
	which only contains reads mapped onto the beginnings of 1Mbp of chromosome 2R.
      </p>

      <h3>Change coordinate of reads according to the gene sequence mapped result</h3>
      <p>Combine gene sequence mapped result against reference sequence and read mapped result against target genes/genomes. 
	<pre>
	  python recot_combine.py -c settings_sample.ini
	  python recot_convert.py -c settings_sample.ini
	</pre>
	This step may require a few hours (depend on number of reads and CPU/Disk speed).
      </p>
      <p>
      <img alt="Transform coordinate" src="images/transform.png" />
      </p>

      <h3>Visualize the result</h3>
      <p>When recot_convert.py is successfully finished, the script shows the way to convert the result SAM file into BAM file.
	<pre>
	  samtools view -bt sample/dmel-2R.fasta -o your_read.bam sample/ERR020078.on_dme.sam
	  samtools sort your_read.bam sample/ERR020078.on_dme.bam
	  samtools index sample/ERR020078.on_dme.bam
	</pre>
	To visualize the result in IGV, you can use the ERR020078.on_dme.bam (and ERR020078.on_dme.bam.bai).
      </p>
    </div>

    <div class="section">
      <h2>FAQ</h2>
      <h3>Q. Can I change configuration file?</h3>
      <p>By adding option "-c", you can change the file name from "settings.ini".</p>

      <h3>Q. Can I transform coordinate of RNA-seq reads even when I do not have genome sequence of original species?</h3>
      <p>Yes. First, you prepare gene sequences in FASTA format (From RNA-seq reads, you can use 
      transcriptome assembly software such as Trinity and OASIS). 
      Then, by skipping "recot_extract.py" in the above procedure and specify the result file 
      at gene_seq_file parameter in "setting.ini", you can transform coordinate of the reads.</p>
      <h3>Q. I would like to use prepared orthologous gene relationships between two species.</h3>
      <p>recot_rm_overlap.py can handle the orthologous genes. To use the relationships, 
	<ol>
	  <li>Prepare the tab delimited file containing gene-gene relation (sample/relation_dsim_dme.tsv). First column is gene name of original species, and second column is gene name of target species.</li>
	  <li>In setting.ini, you need to change two parameters. 1. change the value of "gene_set" to the file including orthologous genes. 2. set target_gff_file, which contains gene names and their regions of target species with GFF format.</li>
	</ol>
      </p>
      
      <h3>Q. Samtools give up convert of SAM to BAM halfway with some errors. What's wrong?</h3>
      <p>Bugs in mapping software or our scripts sometime cause the problem.
      In the case, please run the following command and then use the "output.sam" instead of input.sam:
      <pre>
	python recot_check_cigar.py input.sam output.sam errorreads.sam
      </pre>
      </p>
      <h3>Q. I have billions of reads. Can I increase the speed of the transformation?</h3>
      <p>If you can use SUN(Oracle) Grid Engine or Torque, you can parallelize the transformation in "recot_convert.py" by using array job.
	First, you need to count the number of chromosomes (contigs) of your original species (Suppose that we have 20 contigs).
	Then, run the following command instead of running "recot_convert.py" (for Sun Grid Engine):
	<pre>
	  qsub -t1-20:1 qsub_convert.sh
	  python recot_samjoin.py
	</pre>
	To change a setting file, you need to add "-c" in the command in qsub_convert.sh and run recot_samjoin.py with -c option.
      </p>
      
    </div>

    <div class="section">
      <h2>License</h2>
      <p>BSD License</p>
    </div>
    <div class="section">
    <div>
      <h2>Contact</h2>
      <p>RECOT Development Team:
	<ul>
	  <li>Akiko Izawa (izawa.akiko_AT_sel.is.ocha.ac.jp)</li>
	  <li>Jun Sese (sesejun_AT_cs.titech.ac.jp)</li>
	</ul>
      </p>
    </div>
    <div>
      <h2>Download</h2>
      <p>
	You can download this project in either
	<a href="http://github.com/sesejun/recot/zipball/master">zip</a> or
	<a href="http://github.com/sesejun/recot/tarball/master">tar</a> formats.
      </p>
      <p>You can also clone the project with <a href="http://git-scm.com">Git</a>
	by running:
	<pre>
	  git clone git://github.com/sesejun/recot
	</pre>
      </p>
    </div>
  </div>
  <div class="footer">
    get the source code on GitHub : <a href="http://github.com/sesejun/recot">sesejun/recot</a>
  </div>

</body>
</html>
